cloud/scope/machine.go

```diff
@@ -766,6 +771,99 @@ func (m *MachineScope) DeleteVPCLoadBalancerPoolMember() error {
+
+		// Retrieve the Load Balancer.
+		var loadBalancerDetails *vpcv1.LoadBalancer
+		if member.LoadBalancer.ID != nil {
+			loadBalancerDetails, _, err = m.IBMVPCClient.GetLoadBalancer(&vpcv1.GetLoadBalancerOptions{
+				ID: member.LoadBalancer.ID,
+			})
+		} else if member.LoadBalancer.Name != nil {
+			loadBalancerDetails, err = m.IBMVPCClient.GetLoadBalancerByName(*member.LoadBalancer.Name)
+		} else {
+			return fmt.Errorf("error load balancer has no id or name for load balancer pool member deleteion")
+		}
+		if err != nil {
+			return fmt.Errorf("error retrieving load balancer for load balancer pool member deletion for machine %s: %w", m.IBMVPCMachine.Name, err)
+		}
+		m.Logger.V(5).Info("collected load balancer for load balancer pool member deletion", "machineName", m.IBMVPCMachine.Name, "loadBalancerID", *loadBalancerDetails.ID)
+
+		// Lookup the Load Balancer Pool ID, if only name is available.
+		loadBalancerPoolID, err := m.getLoadBalancerPoolID(ptr.To(member.Pool), *loadBalancerDetails.ID)
+		if err != nil {
+			return fmt.Errorf("error retrieving load balancer pool id for load balancer pool member deletion for machine %s: %w", m.IBMVPCMachine.Name, err)
+		}
+		m.Logger.V(5).Info("collected load balancer pool id for load balancer pool member deletion", "machineName", m.IBMVPCMachine.Name, "loadBalancerPoolID", *loadBalancerPoolID)
+
+		listMembersOptions := &vpcv1.ListLoadBalancerPoolMembersOptions{
+			LoadBalancerID: loadBalancerDetails.ID,
+			PoolID:         loadBalancerPoolID,
+		}
+
+		m.Logger.V(5).Info("list load balancer pool members options", "machineName", m.IBMVPCMachine.Name, "options", *listMembersOptions)
+		poolMembers, _, err := m.IBMVPCClient.ListLoadBalancerPoolMembers(listMembersOptions)
+		if err != nil {
+			return fmt.Errorf("error retrieving load balancer pool members for load balancer pool member deletion for machine %s: %w", m.IBMVPCMachine.Name, err)
+		}
+
+		for _, poolMember := range poolMembers.Members {
+			poolMemberTarget, ok := poolMember.Target.(*vpcv1.LoadBalancerPoolMemberTarget)
+			if !ok {
+				continue
+			}
+			if poolMemberTarget.Address != nil && *poolMemberTarget.Address == *instanceDetails.PrimaryNetworkInterface.PrimaryIP.Address {
+				m.Logger.V(3).Info("found load balancer pool member to delete", "machineName", m.IBMVPCMachine.Name, "poolMemberID", *poolMember.ID)
+				// Make LB status check now that it has been determined a change is required.
+				if *loadBalancerDetails.ProvisioningStatus != string(infrav1beta2.VPCLoadBalancerStateActive) {
+					m.Logger.V(5).Info("load balancer not in active status prior to load balancer pool member deletion", "machineName", m.IBMVPCMachine.Name, "loadBalancerID", *loadBalancerDetails.ID, "loadBalancerProvisioningStatus", *loadBalancerDetails.ProvisioningStatus)
+					// Set flag that some cleanup was not completed, and break out of member target loop, to try next member from Machine Status.
+					cleanupIncomplete = true
```

Comment:
I think we can directly return with error from here right? Anyhow we are not doing anything outside for loop apart from checking `cleanupIncomplete`

Reply:
It is more ideal to attempt to reconcile additional LB members if possible, not block on a single LB at a time.

Correct, once going through all supplied LB Pool Member "requests", if any failed, return the failure. If none failed or were necessary, return nil.

Reply:
In that case of iterating over other pool members, we should use continue instead of break right?

Reply:
No, if we found there is a Pool Member that matches the Machine's IP, that should be removed, but the LB is not active (cannot be updated), we don't need to check any additional Pool Members. So, break from the inner for loop, and continue with the outer for loop, checking the defined Pool Members in Machine Status (if it has more than one Pool Member defined for the Machine).
(Continue would simply continue iterating through the inner for loop, which shouldn't find any additional IP Address matches, since we already found the one matching the Machine IP Address). Shortcut those remaining non-matching members in the inner loop by `break`.

If we do not find a matching Member inside the inner for loop, we don't care what state the LB is in (active or not), so it should simply iterate through the inner loop, finding no matches, and naturally continue the outer for loop, if necessary.

Reply:
Got the point thanks for the explaination, Another question is we are fetching the lb details in the outerloop can't we check teh lb status there instead of checking it inside the second loop?

Reply:
I could, but if we find out the LB Pool Member is already gone (it does not show up in the inner loop), we don't need to worry about the LB status.
Thus, we don't need to wait for every LB to be active, just to check if it has a matching Pool member.
So, if another machine is requesting removal of a member in the same LB, we aren't blocked waiting, just to find out that LB no longer have the member anyway.